feat(har): record remote IP:PORT and SSL details

rwoll opened the pull request:
Updates in the HAR capture Log Entries:

* `serverIPAddress`: (string) This now gets populated.
* `_serverPort`: (number) This now gets populated. There isn't an
  official field according to this [w3 doc][w3] (and I didn't see
  a suitable field in a HAR download from Chromium DevTools).

  I used the field name from [WK WebInspector HARBuilder.js][wk].
* `_securityDetails`: non-standard info that is helpful in making
  assertions about cert expiries.

This also introduces new APIs on `network.Response` exposing the
new info.

NB: At first these APIs were synchronous and populated with info that is
readily available at the beginning of the Response creation, but WK does
not emit IP address info until _loadingFinished, so the API's were all
turned into Promise-based approaches.

Closes #6624.

[w3]: https://www.w3.org/community/bigdata-tools/files/2017/10/HAR_Spec_TO_HAR_Vocabulary.pdf
[wk]: https://opensource.apple.com/source/WebInspectorUI/WebInspectorUI-7608.4.9.1.3/UserInterface/Controllers/HARBuilder.js.auto.html

rwoll committed:
feat(har): record remote IP:PORT and SSL details

Updates in the HAR capture Log Entries:

* `serverIPAddress`: (string) This now gets populated.
* `_serverPort`: (number) This now gets populated. There isn't an
  official field according to this [w3 doc][w3] (and I didn't see
  a suitable field in a HAR download from Chromium DevTools).

  I used the field name from [WK WebInspector HARBuilder.js][wk].
* _securityDetails: non-standard info that is helpful in making
  assertions about cert expiries.

This also introduces new APIs on `network.Response` exposing the
new info.

NB: At first these APIs were synchonous and populated with info that is
readily available at the beginning of the Response creation, but WK does
not emit IP address info until _loadingFinished, so the API's were all
turned into Promise-based approaches.

Closes #6624.

[w3]: https://www.w3.org/community/bigdata-tools/files/2017/10/HAR_Spec_TO_HAR_Vocabulary.pdf
[wk]: https://opensource.apple.com/source/WebInspectorUI/WebInspectorUI-7608.4.9.1.3/UserInterface/Controllers/HARBuilder.js.auto.html

rwoll commented on src/server/webkit/wkInterceptableRequest.ts:
```diff
@@ -126,7 +126,24 @@ export class WKInterceptableRequest implements network.RouteDelegate {
+      securityDetails = {
+        protocol: responsePayload.security.connection?.protocol || '',
+        subjectName: responsePayload.security.certificate?.subject || '',
+        issuer: '',
```
WK is the only browser to not have this. It also appears protocol is always blank for WK. I'm happy to drop everything except `validFrom`/`validTo`. Alternatively, I'm happy to pass everything through and `any` type sec details allowing it to be browser specific. (Though that did not feel in the spirit of PW.)

rwoll commented on src/server/webkit/wkInterceptableRequest.ts:
```diff
@@ -126,7 +126,24 @@ export class WKInterceptableRequest implements network.RouteDelegate {
     };
-    return new network.Response(this.request, responsePayload.status, responsePayload.statusText, headersObjectToArray(responsePayload.headers), timing, getResponseBody);
+
+    const serverIPAddressAndPort: network.IPAddressAndPort = {
+      ipAddress: '',
+      port: -1,
```
TODO: This will be actual values once https://github.com/microsoft/playwright/pull/6627 lands.

rwoll reviewed (commented):
/cc @JoelEinbinder 🎉

rwoll commented on src/server/helper.ts:
```diff
@@ -116,6 +116,29 @@ class Helper {
     return '\n' + '='.repeat(20) + ' Browser output: ' + '='.repeat(20) + '\n' + logs.join('\n');
   }
+
+  static parseRemoteAddress(value?: string) {
+    if (!value)
+      return;
+
+    const colon = value.lastIndexOf(':');
+    const dot = value.lastIndexOf('.');
+    try {
+      if (colon > dot) { // IPv4
+        const [address, port] = value.split(':');
+        return {
+          ipAddress: address,
+          port: +port,
+        };
+      } else { // IPv6
+        const [address, port] = value.split('.');
+        return {
+          ipAddress: `[${address}]`,
+          port: +port,
+        };
+      }
+    } catch (_) {}
+  }
```
Is this too naïve? FF and CR give us the info split up with the proper typing, WK we have to parse it out ourselves.

rwoll commented on src/server/supplements/har/harTracer.ts:
```diff
@@ -212,6 +212,15 @@ export class HarTracer {
     };
     harEntry.time = [dns, connect, ssl, wait, receive].reduce((pre, cur) => cur > 0 ? cur + pre : pre, 0);
+
+    const promise = response.finished().then(() => {
+      const serverIPAddressAndPort = response.serverIPAddressAndPort();
+      harEntry.serverIPAddress = serverIPAddressAndPort.ipAddress === '' ? undefined : serverIPAddressAndPort.ipAddress;
+      harEntry.connection = serverIPAddressAndPort.port === -1 ? undefined : serverIPAddressAndPort.port.toString();
+      harEntry.securityDetails = response.securityDetails();
+    }).catch(() => {});
+    this._addBarrier(page, promise);
+
```
WK only gives us the IP info later in the request lifecycle. Should we change `response.serverIPAddressAndPort()` and `response.securityDetails()` to return promises that internally wait on finished? Without that, if you call those methods for WK without waiting for finished(), you'll get the placeholder values.

JoelEinbinder commented on src/server/helper.ts:
```diff
@@ -116,6 +116,29 @@ class Helper {
     return '\n' + '='.repeat(20) + ' Browser output: ' + '='.repeat(20) + '\n' + logs.join('\n');
   }
+
+  static parseRemoteAddress(value?: string) {
```
This method doesn't work with [::1]:80. I think parsing ipv6 addresses is somewhat complicated. I'd do `new URL('http://' + address)`. Note that if the port is empty is actually 80.

rwoll replied:
WK returns addresses like `::1.8910` which explode explode with `new URL('http://' + address)`

JoelEinbinder replied:
oh no :(. ok if its not a valid ip then i guess its fine

rwoll commented on src/server/network.ts:
```diff
@@ -326,6 +370,14 @@ export class Response extends SdkObject {
     return this._timing;
   }
 
+  serverIPAddressAndPort(): IPAddressAndPort {
+    return this._serverIPAddressAndPort;
+  }
+
+  securityDetails(): SecurityDetails | undefined {
+    return this._securityDetails;
+  }
```
These might need to be change to returning Promises to accom WK's request lifecycle. Thoughts?

JoelEinbinder replied:
They shouldn't ever need to be promises. We would instead just delay emitting the response event.

rwoll replied:
Is this the line that is currently triggering the emission of the response event:

https://github.com/microsoft/playwright/blob/475a417dfc29caeb644eb721cbcabd5bb8980636/src/server/webkit/wkPage.ts#L980

I don't think moving that into 

https://github.com/microsoft/playwright/blob/475a417dfc29caeb644eb721cbcabd5bb8980636/src/server/webkit/wkPage.ts#L991

works out but I also am likely misunderstanding the flow of events.

rwoll replied:
Updated per Slack comments. Everything's a Promise!

rwoll commented on src/server/network.ts:
```diff
@@ -275,15 +288,19 @@ export class Response extends SdkObject {
   private _headersMap = new Map<string, string>();
   private _getResponseBodyCallback: GetResponseBodyCallback;
   private _timing: ResourceTiming;
+  private _serverIPAddressAndPort: IPAddressAndPort;
+  private _securityDetails?: SecurityDetails;
```
@JoelEinbinder Should we combine all this info? Or leave it as separate fields? Also, what's the consensus on using  making all members of SecurityDetails/IPAddressAndPort optional instead of using the `-1/''` special values.

JoelEinbinder replied:
id prefer optional instead of special values

rwoll replied:
Done! I got to delete a lot of awkward code :)

rwoll commented on tests/har.spec.ts:
```diff
@@ -281,3 +281,29 @@ it('should have popup requests', async ({ contextFactory, server }, testInfo) =>
   expect(entries[1].response.status).toBe(200);
 });
+
+it.only('should have connection details', async ({ contextFactory, server, browserName, browserMajorVersion }, testInfo) => {
```
I'll remove this only with the next pass :)

rwoll commented:
@JoelEinbinder I cleaned up the types (including deleting some of the awkward code dealing with the non-standard zero values.) I also added a comment to the WebKit address parsing code with some examples from my testing.

I haven't yet figured out the appropriate changes to emit the Response at the correct time after `_onLoadingFinished` where the IP address details are available for WebKit so I left the signal barrier in place in the HAR code. I'll remove that once we sort out dealing Response emission timing (directly or indirectly through guarding the newly added fields behind a Promise api). If we can avoid adding more external Promises to Response, I agree that would be much nicer.

rwoll commented on src/server/supplements/har/harTracer.ts:
```diff
@@ -212,6 +212,15 @@ export class HarTracer {
     };
     harEntry.time = [dns, connect, ssl, wait, receive].reduce((pre, cur) => cur > 0 ? cur + pre : pre, 0);
+
+    const promise = response.finished().then(() => {
+      const serverIPAddressAndPort = response.serverIPAddressAndPort();
+      harEntry.serverIPAddress = serverIPAddressAndPort?.ipAddress;
+      harEntry.connection = serverIPAddressAndPort?.port?.toString();
+      harEntry.securityDetails = response.securityDetails();
+    }).catch(() => {});
+    this._addBarrier(page, promise);
+
```
This code should go away once we sequence the Response emission for WK differently (or use a Promised based API for the new additions.)

This iteration actually caused CR to regress due to what I think is unexpected behavior with CR response.finish(). Repro/Failing Test Case: https://github.com/microsoft/playwright/pull/6653

rwoll commented:
@JoelEinbinder Here's the latest pass! Promise-based APIs because of WK. See https://github.com/microsoft/playwright/pull/6631#issue-646501474 with updated description.

rwoll commented on src/server/network.ts:
```diff
@@ -476,3 +540,15 @@ export function mergeHeaders(headers: (types.HeadersArray | undefined | null)[])
+
+function filterEmpties(obj: { [k: string]: string|number|undefined } | undefined) {
+  if (obj)
+    return Object.entries(obj).filter(([_, v]) => v !== null && v !== undefined && (typeof v === 'number' ? v >= 0 : v !== '')).reduce((o, [prop, val]) => { const merged = (o || {}); merged[prop] = val; return merged; }, undefined as { [k: string]: string|number|undefined }|undefined);
```
This grew organically and should be re-written as a simple for-loop for readability. Will do w/ follow up change.

rwoll commented:
@JoelEinbinder: This should now pass on all the browser+platform variants. (Thanks platform-dependent WK networking! 😉 )

Let me know your feedback and I will make any necessary adjustments.

It's also worth skimming through https://github.com/microsoft/playwright/pull/6695 to what additional information will be coming in next.

mxschmitt commented on src/server/network.ts:
```diff
@@ -263,6 +264,24 @@ export type ResourceTiming = {
   responseStart: number;
 };
 
+export type IPAddressAndPort = {
```
I think we can call it RemoteAddr, at least in the Go world, this commonly get's used for ip+port.

rwoll replied:
👍 I'll address in next changeset incorporating any comments from Joel.

mxschmitt commented on src/server/helper.ts:
```diff
@@ -116,6 +116,73 @@ class Helper {
+   *
+   * NB: They look IPv4 and IPv6's with ports but use an alternative notation.
+   */
+  static parseRemoteAddress(value?: string) {
```
this get's only used in WebKit, so I would move it to the `wkPage.ts`.

mxschmitt commented on src/server/network.ts:
```diff
@@ -326,6 +381,16 @@ export class Response extends SdkObject {
     return this._timing;
   }
 
+  async serverIPAddressAndPort(): Promise<IPAddressAndPort|null> {
```
I would call it remoteAddr there too. But let's wait for e.g. Joels opinion there.

mxschmitt commented on tests/har.spec.ts:
```diff
@@ -281,3 +281,33 @@ it('should have popup requests', async ({ contextFactory, server }, testInfo) =>
+  expect(serverIPAddress).toMatch(/^127\.0\.0\.1|\[::1\]/);
+  expect(port).toBe(httpsServer.PORT);
+  if (browserName === 'webkit' && platform === 'linux')
+    expect(securityDetails).toBeUndefined();
```
Is there a bug in the WebKit linux network stack?

rwoll replied:
TBD—I haven't yet debugged to figure out if it's intended (and simply a shortcoming of Linux WK/libsoup) or if it's just broken.

rwoll replied:
I'm going to try to fix this in WK implementation. At first glance, it looks like it's available via libsoup, just not fully plumbed. FMI [see this Slack thread](https://playwright.slack.com/archives/C0141EZ6ZNH/p1621909162023100?thread_ts=1621906702.021200&cid=C0141EZ6ZNH).

rwoll replied:
See https://github.com/microsoft/playwright/issues/6759.

We'll address this as a follow up, but I don't think it should block this MR.

JoelEinbinder replied:
Instead of baking in the bad behavior here, let's mark this as failing on linux.

```ts
it.fail(browserName === 'webkit' && platform === 'linux');
```

JoelEinbinder commented on tests/har.spec.ts:
```diff
@@ -281,3 +281,33 @@ it('should have popup requests', async ({ contextFactory, server }, testInfo) =>
+  expect(serverIPAddress).toMatch(/^127\.0\.0\.1|\[::1\]/);
+  expect(port).toBe(server.PORT);
+  if (browserName === 'webkit' && platform === 'darwin')
+    expect(securityDetails).toEqual({protocol: 'SSL 2.0'});
```
WebKit on mac thinks that 127.0.0.1 is SSL? This doesn't match what I see in Safari.

![image](https://user-images.githubusercontent.com/4624233/120503602-4c634c80-c378-11eb-87af-77acca23e3da.png)

rwoll replied:
Oops…this should probably be marked failing for now. It looks like we are getting placeholder values (`SSL 2.0` and `TLS_NULL_WITH_NULL_NULL `) over the wire from WebKit protocol:

(scroll over all the way over to `securityConnection` section)

```json
{
  "method": "Target.dispatchMessageFromTarget",
  "params": {
    "targetId": "page-8",
    "message": "{\"method\":\"Network.loadingFinished\",\"params\":{\"requestId\":\"9.1\",\"timestamp\":0.027748466171033215,\"metrics\":{\"protocol\":\"http/1.1\",\"priority\":\"high\",\"remoteAddress\":\"127.0.0.1:8000\",\"connectionIdentifier\":\"093BB7CD-C601-4A3A-B68F-0654B44123B1\",\"requestHeaders\":{\"Accept\":\"text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8\",\"Upgrade-Insecure-Requests\":\"1\",\"Host\":\"127.0.0.1:8000\",\"User-Agent\":\"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.2 Safari/605.1.15\",\"Accept-Language\":\"en-us\",\"Accept-Encoding\":\"gzip, deflate\",\"Connection\":\"keep-alive\"},\"requestHeaderBytesSent\":354,\"requestBodyBytesSent\":0,\"responseHeaderBytesReceived\":184,\"responseBodyBytesReceived\":99,\"responseBodyDecodedSize\":99,\"securityConnection\":{\"protocol\":\"SSL 2.0\",\"cipher\":\"TLS_NULL_WITH_NULL_NULL\"}}}}"
  },
  "browserContextId": "8000000000000002",
  "pageProxyId": "7"
}
```

JoelEinbinder replied:
This code from WebInspector might be relevant:

```js
    function loadedSecurely()
    {
        if (this.urlComponents.scheme !== "https" && this.urlComponents.scheme !== "wss" && this.urlComponents.scheme !== "sftp")
            return false;
        if (isNaN(this._timingData.secureConnectionStart) && !isNaN(this._timingData.connectionStart))
            return false;
        return true;
    }
```

JoelEinbinder commented on src/server/network.ts:
```diff
@@ -275,15 +294,33 @@ export class Response extends SdkObject {
+    this._connectionDetailsPromise = new Promise(f => {
+      this._connectionDetailsFinishedCallback = f;
+    });
+    if (connectionDetails.complete ?? true)
```
Too fancy! This is a sign that something has gone wrong with the design.

rwoll replied:
Hehe! Totally agree, but might need a few pointers here. For some background:

1. Started simple: just passed in Security Details and Connection Details in create response…but WebKit didn't have this information until _loadingFinished
2. …so tried to move everything there…[but Chromium sometimes doesn't always fire _loadingFinished](https://github.com/microsoft/playwright/pull/6653#discussion_r636250024)…
3. …so arrived at this very awkward design that needs some help 😄 ❤️

JoelEinbinder replied:
When chromium doesn't fire loadingFinished, we never dispatch the response event. So I don't think this patch needs to worry about that.

rwoll replied:
Sweet—yeah, I lead myself astray (and to unideal code) by focusing on that test case! Will update.

JoelEinbinder commented on src/server/network.ts:
```diff
@@ -275,15 +294,33 @@ export class Response extends SdkObject {
+  private _connectionDetailsPromise: Promise<ConnectionDetails>;
 
-  constructor(request: Request, status: number, statusText: string, headers: types.HeadersArray, timing: ResourceTiming, getResponseBodyCallback: GetResponseBodyCallback) {
+  constructor(request: Request, status: number, statusText: string, headers: types.HeadersArray, timing: ResourceTiming, getResponseBodyCallback: GetResponseBodyCallback, connectionDetails: ConnectionDetails & {complete?: boolean}) {
```
Let's not pass in the connection details here, and only pass them in fully formed to _requestFinished.

JoelEinbinder commented on src/server/helper.ts:
```diff
@@ -116,6 +116,73 @@ class Helper {
+      return filtered;
+  }
+
+  static mergeUpdates(prev?: { [k: string]: string|number|undefined }, updates?: { [k: string]: string|number|undefined }) {
```
I'm hoping that this method can be removed in the final version of this patch.

JoelEinbinder commented on src/server/network.ts:
```diff
@@ -263,6 +264,24 @@ export type ResourceTiming = {
+    validTo?: number;
+};
+
+export type ConnectionDetails = {
```
serverIPAddressAndPort and securityDetails seem to be largely independent from one another, and being merged together in this struct appears to complicate the code in this file. Maybe they are two separate top level items?

JoelEinbinder commented on src/server/network.ts:
```diff
@@ -263,6 +264,24 @@ export type ResourceTiming = {
+}
+
+export type SecurityDetails = {
+    protocol?: string;
```
Are these all actually optional in the final HAR? I'd expect to always have a protocol.

rwoll replied:
Yes, everything in the Security domain in WK is marked as optional: https://github.com/WebKit/WebKit/blob/8afe31a018b11741abdf9b4d5bb973d7c1d9ff05/Source/JavaScriptCore/inspector/protocol/Security.json

And this holds in practice. See branching in the tests which I arrived at watching bot fails.

rwoll commented:
@JoelEinbinder I took a quick pass to refactor to defer populating the information across all browser impls in `_onRequestFinished`. Please let me know if the general approach looks good. (i.e. save off intermediate request data in a map until we need it later in request finished.) If all looks kosher, I will clean up (i.e. ensure we're deleting out of the map not to leak/bloat mem; additionally, we likely need to do something in some of the other terminal request handlers (i.e. loading failed)—I'll add some more test cases, too.)

Unfortunately, this brings me back to why I ended up implementing this and then needing to change to the hybrid approach from the last review: _should include postData_ and _should include binary postData_ now hang because we never fire request finished. Do you have tips to ensure we don't fail on these?

rwoll commented on src/server/webkit/wkInterceptableRequest.ts:
```diff
@@ -126,6 +126,7 @@ export class WKInterceptableRequest implements network.RouteDelegate {
       requestStart: timingPayload ? wkMillisToRoundishMillis(timingPayload.requestStart) : -1,
       responseStart: timingPayload ? wkMillisToRoundishMillis(timingPayload.responseStart) : -1,
     };
+
```
There are a number of errant extra newlines that I'll clean up. They are an artifact of refactoring within this MR.

rwoll commented:
> @JoelEinbinder I took a quick pass to refactor to defer populating the information across all browser impls in `_onRequestFinished`. Please let me know if the general approach looks good. (i.e. save off intermediate request data in a map until we need it later in request finished.) If all looks kosher, I will clean up (i.e. ensure we're deleting out of the map not to leak/bloat mem; additionally, we likely need to do something in some of the other terminal request handlers (i.e. loading failed)—I'll add some more test cases, too.)
> 
> Unfortunately, this brings me back to why I ended up implementing this and then needing to change to the hybrid approach you saw in the last review: _should include postData_ and _should include binary postData_ now hang because we never fire request finished. Do you have tips to ensure we don't fail on these? (relates: [#6631 (comment)](https://github.com/microsoft/playwright/pull/6631#discussion_r644123722))

I suppose one approach would be to add `response._connectionDetailsFinished(serverIPAddressAndPort, securityDetails)` that FF and CR can call immediately after `createResponse`, and WK can call this in `_requestFinished(…)`.

rwoll commented:
> > Unfortunately, this brings me back to why I ended up implementing this and then needing to change to the hybrid approach you saw in the last review: _should include postData_ and _should include binary postData_ now hang because we never fire request finished. Do you have tips to ensure we don't fail on these? (relates: [#6631 (comment)](https://github.com/microsoft/playwright/pull/6631#discussion_r644123722))
> 
> I suppose one approach would be to add `response._connectionDetailsFinished(serverIPAddressAndPort, securityDetails)` that FF and CR can call immediately after `createResponse`, and WK can call this [before] `_requestFinished(…)`.

Here's what that might look like: https://github.com/rwoll/playwright/pull/2. @JoelEinbinder Let me know which approach you think is cleaner. I like doing everything in request finished for simplicity/unity, but I'm not sure how to compromise between Chrome and WebKit. I don't love https://github.com/rwoll/playwright/pull/2 since—while it's simple for Chrome and FF—you can accidentally hang the HAR capture in WK if the finish callbacks aren't invoked in all terminal response paths.

rwoll commented:
Updated with option 2 since fixing Chromium not firing loading finished is out of scope.

JoelEinbinder commented on src/server/network.ts:
```diff
@@ -287,12 +305,26 @@ export class Response extends SdkObject {
   }
 
+  _serverIPAddressAndPortFinished(serverIPAddress?: IPAddressAndPort) {
+    this._serverIPAddressAndPortPromiseCallback(helper.filterEmpties(serverIPAddress));
```
I don't think you need filterEmpties here and below. Everything is eventually going to be JSON.stringified into the har.

rwoll replied:
I've updated this, but we will get some empty objects for securityDetails in the HAR. (Tests updated accordingly.)

JoelEinbinder commented on src/server/network.ts:
```diff
@@ -263,6 +264,19 @@ export type ResourceTiming = {
 };
 
+export type IPAddressAndPort = {
+  ipAddress?: string;
```
We can have an ipAddress without a port? Or a port without an ip? Or maybe these can be required and the whole object becomes optional?

Not important to me either way, I just want to make sure I understand. ip without a port hurts my head.

rwoll replied:
Good point. I don't think I've seen a case where we only get one, so will make them both non-optional.

JoelEinbinder commented on src/server/supplements/har/harTracer.ts:
```diff
@@ -212,6 +212,22 @@ export class HarTracer {
     };
     harEntry.time = [dns, connect, ssl, wait, receive].reduce((pre, cur) => cur > 0 ? cur + pre : pre, 0);
+
+    const promise = Promise.all([
+      response.serverIPAddressAndPort().then(server => {
+        if (server?.ipAddress)
+          harEntry.serverIPAddress = server.ipAddress;
+
+        if (server?.port)
+          harEntry._serverPort = server.port;
+      }),
+      response.securityDetails().then(details => {
+        if (details)
+          harEntry._securityDetails = details;
+      }),
+    ]).then(() => {});
+    this._addBarrier(page, promise);
```
I refactored it a bit while reading. Feel free to keep your code if you like it. However if it's functionally different to mine let me know because it means I misunderstood something. Thanks!

```suggestion
    this._addBarrier(page, response.serverIPAddressAndPort().then(server => {
      if (server?.ipAddress)
        harEntry.serverIPAddress = server.ipAddress;
      if (server?.port)
        harEntry._serverPort = server.port;
    }));
    this._addBarrier(page, response.securityDetails().then(details => {
      if (details)
        harEntry._securityDetails = details;
    }));
```

JoelEinbinder commented on src/server/helper.ts:
```diff
@@ -116,6 +116,22 @@ class Helper {
     return '\n' + '='.repeat(20) + ' Browser output: ' + '='.repeat(20) + '\n' + logs.join('\n');
   }
+
+  static filterEmpties(obj: { [k: string]: string|number|undefined } | undefined) {
```
This can go.

mxschmitt commented on src/server/chromium/crNetworkManager.ts:
```diff
@@ -332,6 +347,7 @@ export class CRNetworkManager {
     const response = request.request._existingResponse();
     if (response)
       response._requestFinished(helper.secondsToRoundishMillis(event.timestamp - request._timestamp));
+
```
nit: revert this line-change.

mxschmitt commented on tests/har.spec.ts:
```diff
@@ -281,3 +281,65 @@ it('should have popup requests', async ({ contextFactory, server }, testInfo) =>
+  const { serverIPAddress, _serverPort: port, _securityDetails: securityDetails } = log.entries[0];
+  expect(serverIPAddress).toMatch(/^127\.0\.0\.1|\[::1\]/);
+  expect(port).toBe(server.PORT);
+  expect(securityDetails).toEqual({});
```
should we maybe make it that it returns null if there are no security details / its no secure connection?

rwoll replied:
That's what `filterEmpties` was doing before (well, `undefined` instead of `null`). I'll bring that behavior back without adding back `filterEmpties`.

mxschmitt replied:
Hmm, sorry my bad. I was confused since I thought its a method which users will consume like PPTR [has](https://pptr.dev/#?product=Puppeteer&version=v10.0.0&show=api-httpresponsesecuritydetails). Feel free to do any now what is compliant with the spec.

rwoll replied:
This section does not have a spec (or much precedent) to my knowledge, so it comes down to what we prefer. I've put it back to the empty object `{}` since that simplifies our code and saves you from doing an extra null/undefined check when consuming the HAR. (This is a non-issue in JS/TS with the details?.foo notation, but for Python it's nice to assume the object is there and just check for a property. ie. just `details.get("foo")` and not `(deatils || {}).get("foo")`)

This does, however mean, there will be a bunch of empty objects in the har for non-secure connections.

/cc @JoelEinbinder

mxschmitt reviewed (approved):
Awesome work! (don't forget to rebase)

rwoll commented:
All feedback addressed! Thanks for the reviews @JoelEinbinder and @mxschmitt!

rwoll committed:
test(har): fix failing win32 test

rwoll commented on tests/har.spec.ts:
```diff
@@ -322,7 +322,7 @@ it('should have security details', async ({ contextFactory, httpsServer, browser
   expect(port).toBe(httpsServer.PORT);
   if (browserName === 'webkit' && platform === 'win32')
-    expect(securityDetails).toEqual({subjectName: 'puppeteer-tests', validFrom: 1550084863});
+    expect(securityDetails).toEqual({subjectName: 'puppeteer-tests', validFrom: 1550084863, validTo: -1});
```
Not sure what the best approach is here…should I just mark this test as fully failing on Win32? `filterEmpties` used to filter out the -1's we were getting from WK.
